[import][inductor] Simplify grid handling

jansel opened the pull request:
Before this PR, calling a triton kernel would look like:
```py
kernel.run(a, b, xnumel, grid=grid(xnumel), stream=stream0)
```
where the `grid=` was passed as a callable (function closure) arg.  This PR removes the grid arg:
```py
kernel.run(a, b, xnumel, stream=stream0)
```
instead now the grid computation is included in the kernel launcher, with something like:
```py
def launcher(in_ptr0, out_ptr0, xnumel, stream):
    grid_0 = ((xnumel + 1023) >> 10)
    grid_1 = 1
    grid_2 = 1
    runner(grid_0, grid_1, grid_2, stream, function, metadata, None, launch_enter_hook, launch_exit_hook, in_ptr0, out_ptr0, xnumel)
```

This should be faster, since we remove multiple function/dict calls and are able to specialize the grid computation for each `triton.Config`.

It also allows us to unify the handling of grids between the Python and C++ wrapper code.  Before this, C++ wrapper code didn't actually support dynamic grid sizes and instead burned in a static grid.

This unification allows this PR to be a net deletion of code.

Note the attached diff contains some minor fbcode-only changes.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

jamesjwu commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -1189,18 +1193,19 @@ def make_launcher(self) -> LauncherType:
+
+        grid = GridExpr.from_meta(self.inductor_meta, cfg)
+        lines = [
+            f"def launcher({', '.join(def_args)}, stream):",
```
nit: A comment that shows what these lines actually look like for a sample grid would be helpful, it's a bit hard to follow.

Maybe it'd be easier if grid just had a `gen_code()` function that generated the prefix followed by grid_0 = .. lines?

jansel replied:
I included an example in the PR description.  Full lines are harder to port between Python/C++/eval.

jamesjwu commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -1189,18 +1193,19 @@ def make_launcher(self) -> LauncherType:
+            *[f"    {line}" for line in grid.prefix],
+            f"    grid_0 = {grid.x_grid}",
+            f"    grid_1 = {grid.y_grid}",
+            f"    grid_2 = {grid.z_grid}",
```
To confirm my understanding here, if we passed in a Grid2D or Grid1D based on the inductor meta,  y_grid and z_grid will default to 1 here, correct?

jansel replied:
Yes correct, they will be 1.

jamesjwu commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2122,6 +2127,19 @@ def _pop_config_kwargs(config: dict[str, Any]) -> dict[str, Any]:
+    return {
+        **config.kwargs,
+        "num_warps": config.num_warps,
+        "num_stages": config.num_stages,
```
what happened to the other few fields from Config that are handled by `_pop_config_kwargs`?  (`num_ctas`, `maxnreg`)?

jansel replied:
I excluded them since it will break BC with older versions of Triton.  We should add them at some point, but for a refactor PR I'm trying to limit changes.

jamesjwu reviewed (commented):
Only reviewed the triton_heuristics.py portion as that's what I'm most familiar with, but that part looks good to me: it may be possible to completely replace the python launcher portion with the static launcher I'm working on, though I'd probably have to reimplement all the Grid* classes in C++ to do so

shunting314 commented:
Great to see that numel args are only passed in once for each kernel call rather than once for the kernel call itself and once again for the grid_fn factory call.

Just a few questions in general
1. this is conceptually like INLINE the grid_fn call into the launcher function?
2. do we have some benchmarking results regarding how much overhead is saved by doing this?

jansel commented:
> this is conceptually like INLINE the grid_fn call into the launcher function?

Yes

> do we have some benchmarking results regarding how much overhead is saved by doing this?

I did a benchmark run on the ghstack version of the PR, the overall speedup is small/neutral -- though it does help compile times for cpp_wrapper.

eellison commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1609,21 +1566,26 @@ def define_user_defined_triton_kernel(
         )
+        from .triton import gen_common_triton_imports, TritonKernel
 
+        patch_triton_dtype_repr()
```
I know this is existing code.. but the patch landed a while ago

jansel replied:
Let's remove this in a different PR in case this break something.

eellison reviewed (commented):
nice cleanup ! have a few comments. would you do a full dashboard run with cpp wrapper, aoti etc just bc this is a lot of changes ?

jansel commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2175,150 +2190,226 @@ def foreach(triton_meta, num_warps, filename=None, inductor_meta=None):
+        if block is None or block == 1:
+            return numel
+        if isinstance(numel, int) and isinstance(block, int):
+            return ceildiv(numel, block)  # constant fold
```
_constant_fold requires commutative ops, div is not

eellison replied:
yea, discovered that later..

jansel commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2175,150 +2190,226 @@ def foreach(triton_meta, num_warps, filename=None, inductor_meta=None):
+        if isinstance(numel, int) and isinstance(block, int):
+            return ceildiv(numel, block)  # constant fold
+        if isinstance(block, int) and next_power_of_2(block) == block:
+            # convert div to a shift when we have power of 2 block size
```
I benchmarked this, and it doesn't matter, will remove.

jansel commented on torch/_inductor/autotune_process.py:
```diff
@@ -699,16 +697,15 @@ def run_with_workspace():
 
                 # Handle zero initialization if needed
-                if workspace_arg.zero_mode == WorkspaceZeroMode.ZERO_ON_CALL:
+                if workspace_arg.zero_mode != WorkspaceZeroMode.UNINITIALIZED:
```
Yes, this is a unrelated bugfix.

jansel commented on torch/_inductor/select_algorithm.py:
```diff
@@ -2310,5 +2311,27 @@ def realize_inputs(*args):
+        self.kwargs_int = {}
+        self.kwargs_sym = {}
+        params = inspect.signature(fn).parameters
+        for name, fn_sym, fn_int in [
+            ("cdiv", CeilDiv, ceildiv),
+            ("min", sympy.Min, min),
+            ("max", sympy.Max, max),
```
```py
>>> a,b = sympy.sympify("a, b")
>>> max(a, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jansel/conda/envs/pytorch/lib/python3.10/site-packages/sympy/core/relational.py", line 511, in __bool__
    raise TypeError("cannot determine truth value of Relational")
TypeError: cannot determine truth value of Relational
```

For ceildiv it traces the `-(x/-y)` trick which is incorrect with C++ rounding.

jansel commented on torch/_inductor/codegen/cpp_wrapper_gpu.py:
```diff
@@ -645,5 +475,34 @@ def generate_kernel_call(
 
+    @staticmethod
+    def prepare_triton_wrapper_args(
+        call_args: list[Any], arg_types: list[Any]
```
Let's do typing in a different PR, it is hard to do typing on PRs with fbcode changes since mypy doesn't run properly internally.

jansel commented on torch/_inductor/select_algorithm.py:
```diff
@@ -2310,5 +2311,27 @@ def realize_inputs(*args):
 
 
+class SymbolicGridFn:
+    def __init__(self, fn: Callable[..., tuple[Any, Any, Any]]):
```
Let's do typing in a different PR, it is hard to do typing on PRs with fbcode changes since mypy doesn't run properly internally.

jansel commented:
Note: I've applied the changes in https://www.internalfb.com/diff/D69965761 -- but can't export due to a merge conflict.

eellison reviewed (approved):
accepting bc comments are small and it's a lot of code to continually rebase. assuming theres been one dashboard run without issues

jansel commented:
Yeah I did a dashboard run.  Large compile time wins for cpp_wrapper, but otherwise neutral.

etaf commented:
@jansel , may I add ciflow/xpu to evaluate the impact to XPU backend?

etaf commented on torch/_inductor/codegen/cpp_wrapper_gpu.py:
```diff
@@ -1,188 +1,194 @@
+                    prefix.writeline(f"bool {name},")
+                else:
+                    raise ValueError(f"Unexpected arg type {arg_type}")
+            prefix.writeline("cudaStream_t stream_,")
```
Hi, @jansel, the hard code `cudaStream_t` will fail on XPU. Suggest to use `self.device_codegen.cpp_stream_type()`

jansel replied:
Thanks, I fixed this on the internal version but can't export due merge conflict.

jansel committed:
[import][inductor] Simplify grid handling (#147583)

Summary:
Before this PR, calling a triton kernel would look like:
```py
kernel.run(a, b, xnumel, grid=grid(xnumel), stream=stream0)
```
where the `grid=` was passed as a callable (function closure) arg.  This PR removes the grid arg:
```py
kernel.run(a, b, xnumel, stream=stream0)
```
instead now the grid computation is included in the kernel launcher, with something like:
```py
def launcher(in_ptr0, out_ptr0, xnumel, stream):
    grid_0 = ((xnumel + 1023) >> 10)
    grid_1 = 1
    grid_2 = 1
    runner(grid_0, grid_1, grid_2, stream, function, metadata, None, launch_enter_hook, launch_exit_hook, in_ptr0, out_ptr0, xnumel)
```

This should be faster, since we remove multiple function/dict calls and are able to specialize the grid computation for each `triton.Config`.

It also allows us to unify the handling of grids between the Python and C++ wrapper code.  Before this, C++ wrapper code didn't actually support dynamic grid sizes and instead burned in a static grid.

This unification allows this PR to be a net deletion of code.

cc jeffdaily sunway513 jithunnair-amd pruthvistony ROCmSupport dllehr-amd jataylo hongxiayang naromero77amd albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Pull Request resolved: https://github.com/pytorch/pytorch/pull/147583

Reviewed By: eellison, shunting314

Differential Revision: D69965761

Pulled By: jansel
